google-cloud-firestore/src/main/java/com/google/cloud/firestore/FirestoreOptions.java

```diff
@@ -311,16 +318,31 @@ protected FirestoreOptions(Builder builder) {
       this.openTelemetryOptions = builder.openTelemetryOptions;
     }
 
+    this.traceUtil = com.google.cloud.firestore.telemetry.TraceUtil.getInstance(this);
+
     this.databaseId =
         builder.databaseId != null
             ? builder.databaseId
             : FirestoreDefaults.INSTANCE.getDatabaseId();
 
-    this.channelProvider =
-        builder.channelProvider != null
-            ? builder.channelProvider
-            : GrpcTransportOptions.setUpChannelProvider(
+    if (builder.channelProvider == null) {
+      ApiFunction<ManagedChannelBuilder, ManagedChannelBuilder> channelConfigurator =
+          this.traceUtil.getChannelConfigurator();
+      if (channelConfigurator == null) {
+        this.channelProvider =
+            GrpcTransportOptions.setUpChannelProvider(
                 FirestoreSettings.defaultGrpcTransportProviderBuilder(), this);
+      } else {
+        // Intercept the grpc channel calls to add telemetry info.
+        this.channelProvider =
+            GrpcTransportOptions.setUpChannelProvider(
+                FirestoreSettings.defaultGrpcTransportProviderBuilder()
+                    .setChannelConfigurator(channelConfigurator),
+                this);
+      }
+    } else {
+      this.channelProvider = builder.channelProvider;
```

Comment:
Question - users who are already setting channel providers will not get the tracing channel configurator, is that correct? It makes sense that the user set channel provider takes precedence, but maybe we want to note somewhere in the docs that if you are setting your own channel provider, here's a code snippet to also enable tracing configuration.

Reply:
Yes, that's right. It's a good idea to document this. I have filed b/330404018 to make sure we mention it in our documentation with code samples.